# DATA CLEANING + EDA (Pandas)

Created By: Angelica Rojas

In [152]:
import pandas as pd
import re
import datetime as dt


## Upload Data

The data for this notebook could be found at this link: https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5 . 

For the purpose of this lesson we will use the CSV file of the data.

In [16]:
df = pd.read_csv("BerkeleyPD_Calls_for_Service.csv.xls")

#what does this do?
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

             CVLEGEND  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

Why did we only want to display the first 5 rows of the dataframe?

What if we wanted to see the size of this dataframe?

In [17]:
# number of rows
len(df)

5617

In [18]:
# shape of df (rows, columns)
df.shape

(5617, 11)

# Part 1: DATA CLEANING

## Column Names

What do all these column names even mean? 

On that same website, BPD offers a narrative pdf file that describes the data they provided. https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5


<img src = "DF_col_desc.png">

## Change Column Name(s)

Why would we want to change the column names?

In [19]:
df = df.rename(columns={'CVLEGEND': 'EVENTDESC'})
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

## Investigating Columns

What is the difference between "Block_Location" and "BLKADDR" columns in the dataframe? From the look of the DF that is displayed it looks as though they are almost the same.

In [20]:
#Let's look at the first value in "Block_Location"
df["Block_Location"][0]

'ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n(37.869363, -122.268028)'

In [21]:
#Let's look at the first value in "BLKADDR"
df["BLKADDR"][0]

'ALLSTON WAY & SHATTUCK AVE'

## Create New Columns

What is new information about "Block_Location" that we can actually use and save?

Let's create new columns for the information we extracted from those values.

In [22]:
coordinates = [["".join(x.split()) for x in re.split(r'[()]',i) if x.strip()][-1] for i in df["Block_Location"]]
#print(coordinates)
#new values
longitude =[["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][-1] for i in coordinates]
latitude = [["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][0] for i in coordinates]

#create new columns for latitude and longitude
df["LATITUDE"] = latitude
df["LONGITUDE"] = longitude

#Check if it worked
df.head()



CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

## Drop Columns

We got all the information we needed from "Block_Location" keeping it would be taking up extra room in our dataframe. 

Let's drop the "Block_Location" from the dataframe.

In [23]:
df = df.drop("Block_Location", axis = 1)

#Check if it dropped
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

We can drop other columns that we do not think would add useful information to our analysis. 

Although we did get this data from the Berkeley PD, let's make sure all values in "City" are "Berkeley". Also, Let's make sure the "State" is "CA" for all values.

In [24]:
df.City.unique()

array(['Berkeley'], dtype=object)

In [25]:
df["State"].unique()

array(['CA'], dtype=object)

We checked all the unique values for columns "State" and "City" and they are the results we wanted, therefore, we do not need those columns anymore. 

Drop the columns listed above.

In [26]:
#drop City and State columns
df = df.drop(["City","State"], axis = 1)

#Check if they dropped
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                          BLKADDR   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2               2500 SHATTUCK AVE  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

## Dealing With Null Values (NaN)

<img src = "null_def.png">

REFERENCE: https://pandas.pydata.org/pandas-docs/stable/missing_data.html

This is a big data set and we can't look through each value one at a time. How can we make sure that there is a value for each category?

In [27]:
df.isnull().sum()

CASENO        0
OFFENSE       0
EVENTDT       0
EVENTTM       0
EVENTDESC     0
CVDOW         0
InDbDate      0
BLKADDR      27
LATITUDE      0
LONGITUDE     0
dtype: int64

Let's look at the rows where "BLKADDR" is a null value. Let's make a temporary sub-dataframe.

In [28]:
null_temp = df[pd.isnull(df['BLKADDR'])]
null_temp

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
53    17036936               DISTURBANCE  06/26/2017 12:00:00 AM   18:24   
104   17090713  THEFT FELONY (OVER $950)  04/09/2017 12:00:00 AM   04:15   
224   17024641             BURGLARY AUTO  05/01/2017 12:00:00 AM   21:00   
235   17046547            VEHICLE STOLEN  08/08/2017 12:00:00 AM   17:00   
291   17053694  THEFT MISD. (UNDER $950)  09/07/2017 12:00:00 AM   17:43   
475   17022572            VEHICLE STOLEN  04/22/2017 12:00:00 AM   21:00   
534   17026854      BURGLARY RESIDENTIAL  05/12/2017 12:00:00 AM   09:00   
1228  17091147             BURGLARY AUTO  06/14/2017 12:00:00 AM   03:00   
1306  17020446            VEHICLE STOLEN  04/12/2017 12:00:00 AM   18:00   
1311  17025351           THEFT FROM AUTO  05/04/2017 12:00:00 AM   22:30   
1550  17029419      BURGLARY RESIDENTIAL  05/23/2017 12:00:00 AM   16:30   
1624  17033574                 NARCOTICS  06/11/2017 12:00:00 AM   16:53   
2093  17037072             BURGLARY AUTO  06/20/2017 12:00:00 AM   08:00   
2340  17028899           ALCOHOL OFFENSE  05/21/2017 12:00:00 AM   22:00   
2595  17030776      BURGLARY RESIDENTIAL  05/30/2017 12:00:00 AM   10:45   
3453  17038382      BURGLARY RESIDENTIAL  06/30/2017 12:00:00 AM   15:00   
3922  17034863               DISTURBANCE  06/01/2017 12:00:00 AM   00:00   
4074  17091632                 VANDALISM  08/15/2017 12:00:00 AM   23:30   
4505  17091900  THEFT MISD. (UNDER $950)  09/20/2017 12:00:00 AM   21:35   
4545  17026051           THEFT FROM AUTO  05/07/2017 12:00:00 AM   16:00   
4759  17091711             BURGLARY AUTO  08/28/2017 12:00:00 AM   08:20   
4852  17042511             BURGLARY AUTO  07/20/2017 12:00:00 AM   16:00   
4959  17090810             BURGLARY AUTO  04/23/2017 12:00:00 AM   00:00   
5032  17030164            VEHICLE STOLEN  05/26/2017 12:00:00 AM   20:00   
5194  17091379             BURGLARY AUTO  07/14/2017 12:00:00 AM   12:00   
5406  17020088     ASSAULT/BATTERY MISD.  04/11/2017 12:00:00 AM   09:25   
5544  17091287                 VANDALISM  07/01/2017 12:00:00 AM   08:00   

                   EVENTDESC  CVDOW                InDbDate BLKADDR  \
53        DISORDERLY CONDUCT      1  09/25/2017 03:30:16 AM     NaN   
104                  LARCENY      0  09/25/2017 03:30:12 AM     NaN   
224       BURGLARY - VEHICLE      1  09/25/2017 03:30:12 AM     NaN   
235      MOTOR VEHICLE THEFT      2  09/25/2017 03:30:18 AM     NaN   
291                  LARCENY      4  09/25/2017 03:30:19 AM     NaN   
475      MOTOR VEHICLE THEFT      6  09/25/2017 03:30:12 AM     NaN   
534   BURGLARY - RESIDENTIAL      5  09/25/2017 03:30:12 AM     NaN   
1228      BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM     NaN   
1306     MOTOR VEHICLE THEFT      3  09/25/2017 03:30:12 AM     NaN   
1311  LARCENY - FROM VEHICLE      4  09/25/2017 03:30:12 AM     NaN   
1550  BURGLARY - RESIDENTIAL      2  09/25/2017 03:30:13 AM     NaN   
1624          DRUG VIOLATION      0  09/25/2017 03:30:14 AM     NaN   
2093      BURGLARY - VEHICLE      2  09/25/2017 03:30:16 AM     NaN   
2340    LIQUOR LAW VIOLATION      0  09/25/2017 03:30:13 AM     NaN   
2595  BURGLARY - RESIDENTIAL      2  09/25/2017 03:30:14 AM     NaN   
3453  BURGLARY - RESIDENTIAL      5  09/25/2017 03:30:17 AM     NaN   
3922      DISORDERLY CONDUCT      4  09/25/2017 03:30:16 AM     NaN   
4074               VANDALISM      2  09/25/2017 03:30:18 AM     NaN   
4505                 LARCENY      3  09/25/2017 03:30:19 AM     NaN   
4545  LARCENY - FROM VEHICLE      0  09/25/2017 03:30:12 AM     NaN   
4759      BURGLARY - VEHICLE      1  09/25/2017 03:30:18 AM     NaN   
4852      BURGLARY - VEHICLE      4  09/25/2017 03:30:17 AM     NaN   
4959      BURGLARY - VEHICLE      0  09/25/2017 03:30:12 AM     NaN   
5032     MOTOR VEHICLE THEFT      5  09/25/2017 03:30:13 AM     NaN   
5194      BURGLARY - VEHICLE      5  09/25/2017 03:30:17 AM     NaN   
5406                 ASSAULT     

Does the number of rows in the dataframe match the values above?

In [29]:
#get number of rows of new df
len(null_temp)

27

In [87]:
#get unique values of BLKADDR for the Lat/Long combo
#df("LATITUDE","LONGITUDE").unique()
df1=df.groupby(["LATITUDE","LONGITUDE"])[["BLKADDR"]].count()
#df.groupby(["LATITUDE","LONGITUDE"]).CASENO.size().reset_index()

#df.groupby(["LATITUDE","LONGITUDE"]).CASENO.value_counts()

df1.head()

BLKADDR
LATITUDE                  LONGITUDE         
0Berkeley                 CA               1
0UNKNOWNBerkeley          CA               1
100054THSTBerkeley        CA               1
1000Berkeley              CA               1
1001FAIRGROUNDSDRBerkeley CA               1

## Drop Null Values (NaN)

We could essentially go to Google and try to figure out the BLKADDR ourselves, but to remove any problems that may occur while searching let's just drop all the rows that include null values

In [52]:
#drop rows that have null values
df = df.dropna(axis = 0, how = "any")

In [53]:
#now find out how many null values within the df
#What should you see when you run this?
df.isnull().sum()

CASENO       0
OFFENSE      0
EVENTDT      0
EVENTTM      0
EVENTDESC    0
CVDOW        0
InDbDate     0
BLKADDR      0
LATITUDE     0
LONGITUDE    0
dtype: int64

# <font color = "red">Pandas HW 1</font>

### What is the difference between "EVENTDT" and "EVENTTM"? How can we clean our columns to reflect the data that is useful?

HINT: Focus on EVENTDT

In [59]:
# can we clean this and just keep the date
null_temp.EVENTDT



53      06/26/2017 12:00:00 AM
104     04/09/2017 12:00:00 AM
224     05/01/2017 12:00:00 AM
235     08/08/2017 12:00:00 AM
291     09/07/2017 12:00:00 AM
475     04/22/2017 12:00:00 AM
534     05/12/2017 12:00:00 AM
1228    06/14/2017 12:00:00 AM
1306    04/12/2017 12:00:00 AM
1311    05/04/2017 12:00:00 AM
1550    05/23/2017 12:00:00 AM
1624    06/11/2017 12:00:00 AM
2093    06/20/2017 12:00:00 AM
2340    05/21/2017 12:00:00 AM
2595    05/30/2017 12:00:00 AM
3453    06/30/2017 12:00:00 AM
3922    06/01/2017 12:00:00 AM
4074    08/15/2017 12:00:00 AM
4505    09/20/2017 12:00:00 AM
4545    05/07/2017 12:00:00 AM
4759    08/28/2017 12:00:00 AM
4852    07/20/2017 12:00:00 AM
4959    04/23/2017 12:00:00 AM
5032    05/26/2017 12:00:00 AM
5194    07/14/2017 12:00:00 AM
5406    04/11/2017 12:00:00 AM
5544    07/01/2017 12:00:00 AM
Name: EVENTDT, dtype: object

In [72]:
# Your code here
#df=df.drop("DATE1",axis=1)

new_col = pd.to_datetime(df['EVENTDT'], errors='coerce')
idx=3
df.insert(loc=idx,column='DATE1',value=new_col)
print(df.head())
df=df.drop("EVENTDT",axis=1)
df=df.rename(columns={'DATE1': 'EVENTDT'})

df.head()

     CASENO                   OFFENSE                 EVENTDT      DATE1  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM 2017-06-14   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM 2017-09-01   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM 2017-06-10   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM 2017-04-02   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM 2017-06-10   

  EVENTTM           EVENTDESC  CVDOW                InDbDate  \
0   15:15  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1   22:56     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2   10:45             LARCENY      6  09/25/2017 03:30:15 AM   
3   19:30  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4   14:30      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                          BLKADDR   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  37.

CASENO                   OFFENSE    EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO 2017-06-14   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON 2017-09-01   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950) 2017-06-10   10:45             LARCENY   
3  17018444             BURGLARY AUTO 2017-04-02   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS 2017-06-10   14:30      DRUG VIOLATION   

   CVDOW                InDbDate                         BLKADDR   LATITUDE  \
0      3  09/25/2017 03:30:15 AM      ALLSTON WAY & SHATTUCK AVE  37.869363   
1      5  09/25/2017 03:30:18 AM   UNIVERSITY AVENUE & MILVIA ST  37.871884   
2      6  09/25/2017 03:30:15 AM               2500 SHATTUCK AVE  37.863811   
3      0  09/25/2017 03:30:11 AM    DURANT AVENUE & ELLSWORTH ST  37.867221   
4      6  09/25/2017 03:30:14 AM  MILVIA STREET & UNIVERSITY AVE  37.871884   

     LONGITUDE  
0  -122.268028  
1  -122.270752  
2  -122.267412  
3  -122.263531  
4  -122.270752

### Drop the "InDbDate" column

In [73]:
df = df.drop("InDbDate", axis = 1)

#Check that it actually dropped
df.head()

CASENO                   OFFENSE    EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO 2017-06-14   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON 2017-09-01   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950) 2017-06-10   10:45             LARCENY   
3  17018444             BURGLARY AUTO 2017-04-02   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS 2017-06-10   14:30      DRUG VIOLATION   

   CVDOW                         BLKADDR   LATITUDE    LONGITUDE  
0      3      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1      5   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2      6               2500 SHATTUCK AVE  37.863811  -122.267412  
3      0    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4      6  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

# Part 2: EXPLORATORY DATA ANALYSIS

<h3>"Exploratory data analysis or 'EDA' is a <b>critical</b> beginning step in analyzing the data from an experiment.</h3>

<b>Here are the main reasons we use EDA:</b>
<ul>
• detection of mistakes<br><br>
• checking of assumptions<br><br>
• preliminary selection of appropriate models<br><br>
• determining relationships among the explanatory variables, and<br><br>
• assessing the direction and rough size of relationships between explanatory and outcome variables."</ul>
REFERENCE: http://www.stat.cmu.edu/~hseltman/309/Book/chapter4.pdf


## Now what?

We have cleaned our data to the best of our ability based on the initial look. Now let's try to look at the <b>relationships</b> between different values. 

In [74]:
df.head()

CASENO                   OFFENSE    EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO 2017-06-14   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON 2017-09-01   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950) 2017-06-10   10:45             LARCENY   
3  17018444             BURGLARY AUTO 2017-04-02   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS 2017-06-10   14:30      DRUG VIOLATION   

   CVDOW                         BLKADDR   LATITUDE    LONGITUDE  
0      3      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1      5   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2      6               2500 SHATTUCK AVE  37.863811  -122.267412  
3      0    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4      6  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

Let's look at the different types of offenses that were called in. We know that using the .unique() function will return all the unique values in the column, but what if we wanted to also <b>count</b> the different times each unique value appeared?

In [83]:
df.OFFENSE.value_counts()

BURGLARY AUTO               1069
THEFT MISD. (UNDER $950)     864
VANDALISM                    447
DISTURBANCE                  423
NARCOTICS                    339
VEHICLE STOLEN               306
BURGLARY RESIDENTIAL         271
ASSAULT/BATTERY MISD.        263
THEFT FELONY (OVER $950)     253
ROBBERY                      183
IDENTITY THEFT               150
ALCOHOL OFFENSE              141
THEFT FROM AUTO              137
DOMESTIC VIOLENCE            119
BURGLARY COMMERCIAL          111
ASSAULT/BATTERY FEL.         102
FRAUD/FORGERY                101
MISSING ADULT                 66
GUN/WEAPON                    42
2ND RESPONSE                  42
SEXUAL ASSAULT FEL.           32
BRANDISHING                   25
THEFT FROM PERSON             23
MISSING JUVENILE              22
SEXUAL ASSAULT MISD.          21
ARSON                         16
MUNICIPAL CODE                12
VEHICLE RECOVERED              8
VICE                           1
KIDNAPPING                     1
Name: OFFE

In [84]:
df.EVENTDESC.value_counts()

LARCENY                   1140
BURGLARY - VEHICLE        1069
VANDALISM                  447
DISORDERLY CONDUCT         424
ASSAULT                    365
DRUG VIOLATION             339
MOTOR VEHICLE THEFT        306
BURGLARY - RESIDENTIAL     271
FRAUD                      251
ROBBERY                    183
LIQUOR LAW VIOLATION       141
LARCENY - FROM VEHICLE     137
FAMILY OFFENSE             119
BURGLARY - COMMERCIAL      111
MISSING PERSON              88
WEAPONS OFFENSE             67
SEX CRIME                   53
NOISE VIOLATION             42
ARSON                       16
ALL OTHER OFFENSES          12
RECOVERED VEHICLE            8
KIDNAPPING                   1
Name: EVENTDESC, dtype: int64

Why is "LARCENY" a higher occurence in the "EVENTDESC" column, if when we looked into the "OFFENSE" column, "BURGLARY - VEHICLE" is first? Let's look into this a little more.


## GroupBy 

In [85]:
df.groupby("EVENTDESC").OFFENSE.value_counts()

#turn the series into a DF 
df.groupby("EVENTDESC").OFFENSE.value_counts().to_frame()

OFFENSE
EVENTDESC              OFFENSE                          
ALL OTHER OFFENSES     MUNICIPAL CODE                 12
ARSON                  ARSON                          16
ASSAULT                ASSAULT/BATTERY MISD.         263
                       ASSAULT/BATTERY FEL.          102
BURGLARY - COMMERCIAL  BURGLARY COMMERCIAL           111
BURGLARY - RESIDENTIAL BURGLARY RESIDENTIAL          271
BURGLARY - VEHICLE     BURGLARY AUTO                1069
DISORDERLY CONDUCT     DISTURBANCE                   423
                       VICE                            1
DRUG VIOLATION         NARCOTICS                     339
FAMILY OFFENSE         DOMESTIC VIOLENCE             119
FRAUD                  IDENTITY THEFT                150
                       FRAUD/FORGERY                 101
KIDNAPPING             KIDNAPPING                      1
LARCENY                THEFT MISD. (UNDER $950)      864
                       THEFT FELONY (OVER $950)      253
                       THEFT FROM PERSON              23
LARCENY - FROM VEHICLE THEFT FROM AUTO               137
LIQUOR LAW VIOLATION   ALCOHOL OFFENSE               141
MISSING PERSON         MISSING ADULT                  66
                       MISSING JUVENILE               22
MOTOR VEHICLE THEFT    VEHICLE STOLEN                306
NOISE VIOLATION        2ND RESPONSE                   42
RECOVERED VEHICLE      VEHICLE RECOVERED               8
ROBBERY                ROBBERY                       183
SEX CRIME              SEXUAL ASSAULT FEL.            32
                       SEXUAL ASSAULT MISD.           21
VANDALISM              VANDALISM                     447
WEAPONS OFFENSE        GUN/WEAPON                     42
                       BRANDISHING                    25

From this DF we can see that "LARCENY" has the most OFFENSES within that category. When you add the totals from "THEFT MISD ( UNDER $ 950)" ," THEFT  FELONY (OVER $ 950)", and
"THEFT FROM PERSON" they equal more than "BURGLARY - VEHICLE", but "BURGLARY AUTO" as an offense alone is the highest in number.

## More about GROUP BY
"This grouped variable is now a GroupBy object. It has not actually computed anything yet except for some intermediate data about the group key df['key1']. The idea is that this object has all of the information needed to then apply some operation to each of the groups." - Python for Data Analysis

In [86]:
#Use list() to show what a grouping looks like

df.groupby("EVENTDESC")

Descriptive statistics by group

In [87]:
#returns a dict of your groups
df.groupby("EVENTDESC").groups

{'ALL OTHER OFFENSES': Int64Index([333, 577, 1231, 1609, 1615, 3518, 3949, 4133, 4159, 4954, 5085,
             5475],
            dtype='int64'),
 'ARSON': Int64Index([ 714, 1026, 1203, 1419, 1687, 1771, 2047, 2329, 2615, 3471, 4198,
             4463, 4541, 4757, 4914, 5178],
            dtype='int64'),
 'ASSAULT': Int64Index([  16,   36,   38,   77,   81,   84,   94,   97,   98,  107,
             ...
             5397, 5408, 5413, 5415, 5433, 5459, 5527, 5573, 5584, 5606],
            dtype='int64', length=365),
 'BURGLARY - COMMERCIAL': Int64Index([  15,   78,  153,  285,  308,  448,  476,  494,  556,  606,
             ...
             5218, 5267, 5280, 5290, 5294, 5400, 5437, 5440, 5464, 5563],
            dtype='int64', length=111),
 'BURGLARY - RESIDENTIAL': Int64Index([  46,   50,   52,   56,  132,  147,  170,  185,  186,  188,
             ...
             5383, 5458, 5472, 5483, 5485, 5503, 5555, 5566, 5574, 5610],
            dtype='int64', length=271),
 'BURGLARY - VEHICL

In [88]:
df.groupby("EVENTDESC").describe()

CASENO                                          \
                         count          mean           std         min   
EVENTDESC                                                                
ALL OTHER OFFENSES        12.0  1.704152e+07  13213.748627  17020070.0   
ARSON                     16.0  1.703960e+07  10884.435612  17019292.0   
ASSAULT                  365.0  1.703659e+07  11777.817829  17017269.0   
BURGLARY - COMMERCIAL    111.0  1.703836e+07  10705.445304  17017815.0   
BURGLARY - RESIDENTIAL   271.0  1.703699e+07  11114.258235  17017275.0   
BURGLARY - VEHICLE      1069.0  1.706054e+07  27556.587988  17017257.0   
DISORDERLY CONDUCT       424.0  1.703669e+07  13562.175517  17017299.0   
DRUG VIOLATION           339.0  1.703520e+07  11509.294602  17000233.0   
FAMILY OFFENSE           119.0  1.703587e+07  11453.118362  17017700.0   
FRAUD                    251.0  1.705983e+07  27806.834983  17017772.0   
KIDNAPPING                 1.0  1.703128e+07           NaN  17031279.0   
LARCENY                 1140.0  1.705789e+07  28130.541333  17017278.0   
LARCENY - FROM VEHICLE   137.0  1.703589e+07  11332.733189  17017738.0   
LIQUOR LAW VIOLATION     141.0  1.703791e+07  10850.095943  17017407.0   
MISSING PERSON            88.0  1.703696e+07  11343.280160  17017380.0   
MOTOR VEHICLE THEFT      306.0  1.703768e+07  11536.028034  17017428.0   
NOISE VIOLATION           42.0  1.703601e+07  13390.380016  17019460.0   
RECOVERED VEHICLE          8.0  1.704233e+07  12372.109030  17021305.0   
ROBBERY                  183.0  1.703730e+07  11364.782521  17017334.0   
SEX CRIME                 53.0  1.703842e+07  11780.906933  17018130.0   
VANDALISM                447.0  1.706278e+07  28498.975421  17017227.0   
WEAPONS OFFENSE           67.0  1.703511e+07  11530.087936  17017425.0   

                                                                          \
                                25%         50%          75%         max   
EVENTDESC                                                                  
ALL OTHER OFFENSES      17030886.00  17048863.0  17050971.50  17055473.0   
ARSON                   17033709.25  17043089.0  17045809.00  17055501.0   
ASSAULT                 17026159.00  17036388.0  17047029.00  17056953.0   
BURGLARY - COMMERCIAL   17030141.00  17038580.0  17046785.50  17056543.0   
BURGLARY - RESIDENTIAL  17027670.00  17037319.0  17045792.00  17057172.0   
BURGLARY - VEHICLE      17037109.00  17051099.0  17091218.00  17091917.0   
DISORDERLY CONDUCT      17025385.50  17036020.0  17045683.00  17091914.0   
DRUG VIOLATION          17024758.50  17034060.0  17044695.00  17057201.0   
FAMILY OFFENSE          17025843.50  17036348.0  17046194.00  17056597.0   
FRAUD                   17035605.00  17050606.0  17091074.00  17091857.0   
KIDNAPPING              17031279.00  17031279.0  17031279.00  17031279.0   
LARCENY                 17033049.50  17048366.5  17091112.00  17091916.0   
LARCENY - FROM VEHICLE  17024643.00  17036057.0  17045147.00  17057051.0   
LIQUOR LAW VIOLATION    17029254.00  17036775.0  17047894.00  17055981.0   
MISSING PERSON          17026488.25  17036270.5  17046137.25  17056806.0   
MOTOR VEHICLE THEFT     17027409.00  17038263.0  17047354.75  17057115.0   
NOISE VIOLATION         17023729.25  17035985.0  17048747.50  17056027.0   
RECOVERED VEHICLE       17039963.00  17045642.0  17049294.00  17054894.0   
ROBBERY                 17027413.50  17039067.0  17047388.00  17056854.0   
SEX CRIME               17027123.00  17040240.0  17047789.00  17056130.0   
VANDALISM               17035518.00  17054680.0  17091263.50  17091918.0   
WEAPONS OFFENSE         17023529.00  17035137.0  17043745.00  17054027.0   

                         CVDOW                                                
                         count      mean       std  min   25%  50%  75%  max  
EVENTDESC                                                                     
ALL OTHER OFFENSES        12.0  2.5

### Group By multiple columns

In [89]:
eg = df.groupby(["EVENTDESC","OFFENSE"]).count()
eg

CASENO  EVENTDT  EVENTTM  \
EVENTDESC              OFFENSE                                              
ALL OTHER OFFENSES     MUNICIPAL CODE                12       12       12   
ARSON                  ARSON                         16       16       16   
ASSAULT                ASSAULT/BATTERY FEL.         102      102      102   
                       ASSAULT/BATTERY MISD.        263      263      263   
BURGLARY - COMMERCIAL  BURGLARY COMMERCIAL          111      111      111   
BURGLARY - RESIDENTIAL BURGLARY RESIDENTIAL         271      271      271   
BURGLARY - VEHICLE     BURGLARY AUTO               1069     1069     1069   
DISORDERLY CONDUCT     DISTURBANCE                  423      423      423   
                       VICE                           1        1        1   
DRUG VIOLATION         NARCOTICS                    339      339      339   
FAMILY OFFENSE         DOMESTIC VIOLENCE            119      119      119   
FRAUD                  FRAUD/FORGERY                101      101      101   
                       IDENTITY THEFT               150      150      150   
KIDNAPPING             KIDNAPPING                     1        1        1   
LARCENY                THEFT FELONY (OVER $950)     253      253      253   
                       THEFT FROM PERSON             23       23       23   
                       THEFT MISD. (UNDER $950)     864      864      864   
LARCENY - FROM VEHICLE THEFT FROM AUTO              137      137      137   
LIQUOR LAW VIOLATION   ALCOHOL OFFENSE              141      141      141   
MISSING PERSON         MISSING ADULT                 66       66       66   
                       MISSING JUVENILE              22       22       22   
MOTOR VEHICLE THEFT    VEHICLE STOLEN               306      306      306   
NOISE VIOLATION        2ND RESPONSE                  42       42       42   
RECOVERED VEHICLE      VEHICLE RECOVERED              8        8        8   
ROBBERY                ROBBERY                      183      183      183   
SEX CRIME              SEXUAL ASSAULT FEL.           32       32       32   
                       SEXUAL ASSAULT MISD.          21       21       21   
VANDALISM              VANDALISM                    447      447      447   
WEAPONS OFFENSE        BRANDISHING                   25       25       25   
                       GUN/WEAPON                    42       42       42   

                                                 CVDOW  BLKADDR  LATITUDE  \
EVENTDESC              OFFENSE                                              
ALL OTHER OFFENSES     MUNICIPAL CODE               12       12        12   
ARSON                  ARSON                        16       16        16   
ASSAULT                ASSAULT/BATTERY FEL.        102      102       102   
                       ASSAULT/BATTERY MISD.       263      263       263   
BURGLARY - COMMERCIAL  BURGLARY COMMERCIAL         111      111       111   
BURGLARY - RESIDENTIAL BURGLARY RESIDENTIAL        271      271       271   
BURGLARY - VEHICLE     BURGLARY AUTO              1069     1069      1069   
DISORDERLY CONDUCT     DISTURBANCE                 423      423       423   
                       VICE                          1        1         1   
DRUG VIOLATION         NARCOTICS                   339      339       339   
FAMILY OFFENSE         DOMESTIC VIOLENCE           119      119       119   
FRAUD                  FRAUD/FORGERY               101      101       101   
                       IDENTITY THEFT              150      150       150   
KIDNAPPING             KIDNAPPING                    1        1         1   
LARCENY                THEFT FELONY (OVER $950)    253      253       253   
                       THEFT FROM PERSON            23       23        23   
                       THEFT MISD. (UNDER $950)    864      864       864   
LARCENY - FROM VEHICLE THEFT FROM AUTO             137      137       137   
LIQUOR LAW VIOLATION   ALCOHOL OFFENSE      

### VS

In [90]:
df.groupby("EVENTDESC")[["OFFENSE"]].count()

OFFENSE
EVENTDESC                      
ALL OTHER OFFENSES           12
ARSON                        16
ASSAULT                     365
BURGLARY - COMMERCIAL       111
BURGLARY - RESIDENTIAL      271
BURGLARY - VEHICLE         1069
DISORDERLY CONDUCT          424
DRUG VIOLATION              339
FAMILY OFFENSE              119
FRAUD                       251
KIDNAPPING                    1
LARCENY                    1140
LARCENY - FROM VEHICLE      137
LIQUOR LAW VIOLATION        141
MISSING PERSON               88
MOTOR VEHICLE THEFT         306
NOISE VIOLATION              42
RECOVERED VEHICLE             8
ROBBERY                     183
SEX CRIME                    53
VANDALISM                   447
WEAPONS OFFENSE              67

### Get Columns + Index

In [91]:
eg.columns

Index(['CASENO', 'EVENTDT', 'EVENTTM', 'CVDOW', 'BLKADDR', 'LATITUDE',
       'LONGITUDE'],
      dtype='object')

In [92]:
list(df.columns)

['CASENO',
 'OFFENSE',
 'EVENTDT',
 'EVENTTM',
 'EVENTDESC',
 'CVDOW',
 'BLKADDR',
 'LATITUDE',
 'LONGITUDE']

# <font color = "red">Pandas HW 2</font>

Could there be any relationship with the Day of the Week and the calls? Try out different functions to see if there is any significance?

In [118]:
# count the amount of calls per day
# Your code here ...
#number of offenses grouped by Day Of the Week
df.groupby("CVDOW").OFFENSE.count()

CVDOW
0    702
1    793
2    853
3    818
4    797
5    818
6    809
Name: OFFENSE, dtype: int64

With the day that has the most calls, check the type of offense that appears the most.

In [172]:
# Your code here ...
#df.groupby("CVDOW").OFFENSE.value_counts().to_frame()

#The Offense that happened the most on the Day of the Week i.e. 2, when most calls where received

mod=df.groupby("CVDOW").OFFENSE.count().idxmax()
is_dow=df["CVDOW"]==mod
df_dow=df[is_dow]
print(df_dow.shape)
g=df_dow.groupby("OFFENSE").CASENO.count()
g.nlargest(1).to_frame()



(853, 9)


CASENO
OFFENSE              
BURGLARY AUTO     173

# <font color = "red">Pandas HW 3</font>

Do these numbers match the results of the overall DF?


Let's try something else, Friday and Saturday nights are typically associated with being the "party" time. If this is true should there be more Liquor/Drug/Disordely Conduct/etc. occurrences those nights?

Let's try it with Fridays!

In [190]:
#only display rows with the D.O.W of Friday
#create a temp df

# your code here ...
is_friday=df["CVDOW"]==5
df_friday=df[is_friday]
#print(df_friday.shape)

print("Offense that happened most on Friday:")
print(df_friday.OFFENSE.value_counts().nlargest(1).to_frame())
print("\n")
print("Offense that happened least on Friday:")
print(df_friday.groupby("OFFENSE").CASENO.count().nsmallest(1).to_frame())
print("\n")
print("Offenses on Friday:")
print(df_friday.groupby("OFFENSE").CASENO.count())



Offense that happened most on Friday:
               OFFENSE
BURGLARY AUTO      137


Offense that happened least on Friday:
                CASENO
OFFENSE               
MUNICIPAL CODE       1


Offenses on Friday:
OFFENSE
2ND RESPONSE                 13
ALCOHOL OFFENSE              28
ARSON                         2
ASSAULT/BATTERY FEL.          8
ASSAULT/BATTERY MISD.        33
BRANDISHING                   3
BURGLARY AUTO               137
BURGLARY COMMERCIAL          14
BURGLARY RESIDENTIAL         38
DISTURBANCE                  50
DOMESTIC VIOLENCE            18
FRAUD/FORGERY                13
GUN/WEAPON                    9
IDENTITY THEFT               17
MISSING ADULT                13
MISSING JUVENILE              4
MUNICIPAL CODE                1
NARCOTICS                    59
ROBBERY                      16
SEXUAL ASSAULT FEL.           4
SEXUAL ASSAULT MISD.          2
THEFT FELONY (OVER $950)     43
THEFT FROM AUTO              25
THEFT FROM PERSON             8
THEFT MI

# <font color = "red">Pandas HW 4</font>
Let's try it with Saturdays!

In [182]:
#only display rows with the D.O.W of Saturday
#create a temp df

# your code here ...
is_saturday=df["CVDOW"]==6
df_saturday=df[is_saturday]
df_saturday.shape

(809, 9)

In [193]:
#count the number of offenses by type
df_saturday.OFFENSE.value_counts().to_frame()

OFFENSE
THEFT MISD. (UNDER $950)      126
BURGLARY AUTO                 125
NARCOTICS                      64
DISTURBANCE                    63
VANDALISM                      55
BURGLARY RESIDENTIAL           41
THEFT FELONY (OVER $950)       39
VEHICLE STOLEN                 39
ASSAULT/BATTERY MISD.          37
ALCOHOL OFFENSE                35
ASSAULT/BATTERY FEL.           25
2ND RESPONSE                   25
THEFT FROM AUTO                21
ROBBERY                        20
DOMESTIC VIOLENCE              15
IDENTITY THEFT                 15
BURGLARY COMMERCIAL            14
FRAUD/FORGERY                  14
MISSING ADULT                   6
SEXUAL ASSAULT FEL.             6
GUN/WEAPON                      6
MISSING JUVENILE                4
BRANDISHING                     4
SEXUAL ASSAULT MISD.            4
ARSON                           3
VEHICLE RECOVERED               2
MUNICIPAL CODE                  1

In [192]:
#count the number of events by type
df_saturday.EVENTDESC.value_counts().to_frame()

EVENTDESC
LARCENY                       165
BURGLARY - VEHICLE            125
DRUG VIOLATION                 64
DISORDERLY CONDUCT             63
ASSAULT                        62
VANDALISM                      55
BURGLARY - RESIDENTIAL         41
MOTOR VEHICLE THEFT            39
LIQUOR LAW VIOLATION           35
FRAUD                          29
NOISE VIOLATION                25
LARCENY - FROM VEHICLE         21
ROBBERY                        20
FAMILY OFFENSE                 15
BURGLARY - COMMERCIAL          14
WEAPONS OFFENSE                10
MISSING PERSON                 10
SEX CRIME                      10
ARSON                           3
RECOVERED VEHICLE               2
ALL OTHER OFFENSES              1

# <font color = "red"> GROUP WORK</font> 
<i>"Companies hire Data Scientists who can clearly and fluently translate their technical findings to a non-technical team, such as the Marketing or Sales departments. The goal is to enable the business to make decisions by arming them with quantified insights, in addition to understanding the needs of their non-technical colleagues in order to wrangle the data appropriately"</i>

## What do YOU want to find out? YOUR DATA INVESTIGATION


In this notebook you have been learning all these techniques to be able to manipulate your dataframe to your preference. We know how to clean and explore our data, but what questions or topics did you actually want to learn from the data? 

<b> * In groups of 2-4 people, investigate the dataframe in this notebook and pick a question/topic to answer. Using the techniques you learned today, show relationships and results that would support that question/topic. 
</b><br><br>
<b>Each</b> group will present their investigations and why they are significant to the class.
